chips/sam4l/src/adc.rs

```diff
@@ -43,113 +48,118 @@ use scif;
 #[allow(dead_code,missing_copy_implementations)]
 #[cfg_attr(rustfmt, rustfmt_skip)]
 pub struct AdcRegisters { // From page 1005 of SAM4L manual
-    cr:        usize,   // Control               (0x00)
-    cfg:       usize,   // Configuration         (0x04)
-    sr:        usize,   // Status                (0x08)
-    scr:       usize,   // Status clear          (0x0c)
-    pad:       usize,   // padding/reserved
-    seqcfg:    usize,   // Sequencer config      (0x14)
-    cdma:      usize,   // Config DMA            (0x18)
-    tim:       usize,   // Timing config         (0x1c)
-    itimer:    usize,   // Internal timer        (0x20)
-    wcfg:      usize,   // Window config         (0x24)
-    wth:       usize,   // Window threshold      (0x28)
-    lcv:       usize,   // Last converted value  (0x2c)
-    ier:       usize,   // Interrupt enable      (0x30)
-    idr:       usize,   // Interrupt disable     (0x34)
-    imr:       usize,   // Interrupt mask        (0x38)
-    calib:     usize,   // Calibration           (0x3c)
-    version:   usize,   // Version               (0x40)
-    parameter: usize,   // Parameter             (0x44)
+    cr:        VolatileCell<u32>,   // Control               (0x00)
+    cfg:       VolatileCell<u32>,   // Configuration         (0x04)
+    sr:        VolatileCell<u32>,   // Status                (0x08)
+    scr:       VolatileCell<u32>,   // Status clear          (0x0c)
+    pad:       VolatileCell<u32>,   // padding/reserved
+    seqcfg:    VolatileCell<u32>,   // Sequencer config      (0x14)
+    cdma:      VolatileCell<u32>,   // Config DMA            (0x18)
+    tim:       VolatileCell<u32>,   // Timing config         (0x1c)
+    itimer:    VolatileCell<u32>,   // Internal timer        (0x20)
+    wcfg:      VolatileCell<u32>,   // Window config         (0x24)
+    wth:       VolatileCell<u32>,   // Window threshold      (0x28)
+    lcv:       VolatileCell<u32>,   // Last converted value  (0x2c)
+    ier:       VolatileCell<u32>,   // Interrupt enable      (0x30)
+    idr:       VolatileCell<u32>,   // Interrupt disable     (0x34)
+    imr:       VolatileCell<u32>,   // Interrupt mask        (0x38)
+    calib:     VolatileCell<u32>,   // Calibration           (0x3c)
+    version:   VolatileCell<u32>,   // Version               (0x40)
+    parameter: VolatileCell<u32>,   // Parameter             (0x44)
 }
 
 // Page 59 of SAM4L data sheet
-pub const BASE_ADDRESS: usize = 0x40038000;
+const BASE_ADDRESS: *mut AdcRegisters = 0x40038000 as *mut AdcRegisters;
 
 pub struct Adc {
     registers: *mut AdcRegisters,
-    enabled: bool,
+    enabled: Cell<bool>,
     channel: Cell<u8>,
-    request: Cell<Option<&'static Request>>,
+    client: TakeCell<&'static hil::adc::Client>,
```

Comment:
This can be a `Cell` actually, since the reference doesn't need to be mutable. I think we use `TakeCell` in some other places like this because it saves some typing (`.map()` insted of `.get().map()`), but that's probably not a worthwhile trade off.